lib/connection/connection.js

```diff
@@ -411,6 +421,183 @@ function Connection(context)
+    }
+
+    return status;
+  }
+
+  /**
+   * Gets the status of the query based on queryId.
+   *
+   * @param {String} queryId
+   *
+   * @returns {String} the query status.
+   */
+  this.getQueryStatus = async function (queryId) {
+    return extractQueryStatus(await getQueryResponse(queryId));
+  };
+
+  /**
+   * Gets the status of the query based on queryId and throws if there's an error.
+   *
+   * @param {String} queryId
+   *
+   * @returns {String} the query status.
+   */
+  this.getQueryStatusThrowIfError = async function (queryId) {
+    const response = await getQueryResponse(queryId);
+    const queries = response['data']['queries'];
+    const status = extractQueryStatus(response);
+    let message, code, sqlState = null;
+
+    if (this.isAnError(status)) {
+      message = response['message'] || '';
+      code = response['code'] || -1;
+
+      if (response['data']) {
+        message += queries.length > 0 ? queries[0]['errorMessage'] : '';
+        sqlState = response['data']['sqlState'];
+      }
+
+      throw Errors.createOperationFailedError(
+        code, response, message, sqlState);
+    }
+
+    return status;
+  };
+
+  /**
+   * Gets the results from a previously ran query based on queryId
+   *
+   * @param {Object} options
+   *
+   * @returns {Object}
+   */
+  this.getResultsFromQueryId = async function (options) {
+    const queryId = options.queryId;
+    let status, noDataCounter = 0, retryPatternPos = 0;
+
+    // Wait until query has finished executing
+    let queryStillExecuting = true;
+    while (queryStillExecuting) {
+      // Check if query is still running and trigger exception if it failed
+      status = await this.getQueryStatusThrowIfError(queryId);
+      queryStillExecuting = this.isStillRunning(status);
+      if (!queryStillExecuting) {
+        break;
+      }
+
+      // Timeout based on query status retry rules
+      await new Promise((resolve) => {
+        setTimeout(() => resolve(), asyncRetryInMilliseconds * asyncRetryPattern[retryPatternPos]);
+      });
+
+      // If no data, increment the no data counter
+      if (QueryStatus[status] == QueryStatus.NO_DATA) {
+        noDataCounter++;
+        // Check if retry for no data is exceeded
+        if (noDataCounter > asyncNoDataMaxRetry) {
+          throw Errors.createClientError(
+            ErrorCodes.ERR_GET_RESULTS_QUERY_ID_NO_DATA, true, queryId);
+        }
+      }
+
+      if (retryPatternPos < asyncRetryPattern.length - 1) {
+        retryPatternPos++;
+      }
+    }
+
+    if (QueryStatus[status] != QueryStatus.SUCCESS) {
+      throw Errors.createClientError(
+        ErrorCodes.ERR_GET_RESULTS_QUERY_ID_NOT_SUCCESS_STATUS, true, queryId, status);
+    }
+
+    return this.fetchResult(options);
+  };
+
+  /**
+   * Checks whether the given status is currently running.
+   *
+   * @param {String} status
+   *
+   * @returns {Boolean}
+   */
+  this.isStillRunning = function (status) {
+    const stillRunning =
```

Comment:
How about extracting the `stillRunning` and `isAnError` arrays to query_status.js file and maybe renaming them to `runningStatuses` and `errorStatuses`?

Reply:
Sure, that sounds good